packages/libs/chainweb-stream-client/src/index.ts

```diff
@@ -0,0 +1,303 @@
+  }
+
+  /**
+   * Call to initialize connection
+   */
+  public connect = (): void => {
+    this._debug('connect');
+    this._desiredState = ConnectionState.Connected;
+    this._eventSource = new EventSource(this._makeConnectionURL());
+    this._eventSource.onopen = this._handleConnect;
+    this._eventSource.onerror = this._handleError;
+    // reset & set custom connect timeout handler
+    if (this._connectTimer) {
+      clearTimeout(this._connectTimer);
+    }
+    this._connectTimer = setTimeout(
+      () => this._handleError(new ConnectTimeoutError()),
+      this.connectTimeoutMs,
+    );
+  };
+
+  /**
+   * Call to disconnect
+   */
+  public disconnect = (): void => {
+    this._desiredState = ConnectionState.Closed;
+    this._eventSource?.close();
+    // TODO Should we null out this._eventSource?
+    this._stopHeartbeatMonitor();
+    this._debug('disconnect');
+  };
+
+  /**
+   * .state getter - get current stream client state
+   */
+  public get state(): ConnectionState {
+    const { _eventSource } = this;
+    // special case: are we waiting to reconnect? (expo backoff)
+    if (this._desiredState === ConnectionState.WaitReconnect) {
+      return this._desiredState;
+    }
+    if (_eventSource) {
+      return _eventSource.readyState;
+    }
+    return ConnectionState.None;
+  }
+
+  private _handleConnect = (): void => {
+    const {
+      _failedConnectionAttempts: consecutiveFailedAttempts,
+      _totalConnectionAttempts: totalAttempts,
+      _eventSource,
+    } = this;
+    this._debug('_handleConnect', { consecutiveFailedAttempts, totalAttempts });
+    this._failedConnectionAttempts = 0;
+    if (!_eventSource) {
+      throw new Error(
+        'ChainwebStream._handleConnect called without an _eventSource. This should never happen',
+      );
+    }
+    _eventSource.addEventListener('initial', this._handleData);
+    _eventSource.addEventListener('message', this._handleData);
+    _eventSource.addEventListener('ping', this._resetHeartbeatTimeout);
+    this._resetHeartbeatTimeout();
+    if (this._connectTimer) {
+      clearTimeout(this._connectTimer);
+    }
+    this.emit(this._totalConnectionAttempts ? 'reconnect' : 'connect');
+    this._totalConnectionAttempts += 1;
+  };
+
+  /*
+   * Chrome has different reconnecting strategy (some) vs Firefox (none)
+   * so we disable reconnections and handle them manually
+   */
+  private _handleError = (
+    err: HeartbeatTimeoutError | ConnectTimeoutError | Event,
+  ): void => {
+    this._failedConnectionAttempts += 1;
+    this._totalConnectionAttempts += 1;
+
+    // close event source; crucial - chrome reconnects, firefox does not
+    this._eventSource?.close();
+
+    // cancel connection timeout timer, if exists
+    if (this._connectTimer) {
+      clearTimeout(this._connectTimer);
+    }
+
+    let message = 'Connection error'; // default for "Event". No error information is available, presume disconnection
```

Comment:
if you'd extract the message retrieval you could remove the `let` and then this code would be easier to read

Reply:
good point, I refactored the errors a bit. Please check again